### ISSUE TITLE
Question regarding model usage - False Positive Issues

### ISSUE THREAD

**Issue by laurenzheidrich** [microsoft/BiomedParse#21]
Dear authors,

Thank you for making this repository public!

I am starting off by playing around with your code, especially with the inference_colab_demo.ipynb. I am trying to run inference on some of your example images as well as on some other images.

What I am noticing is that pretty much no matter what the text prompt is, the predicted segmentation is always the same, especially when using prompts of organs that are not present in the scan.

Here are some examples:

For the CT lung scan in your examples, I prompt "covid-19 infection", I get:

![Image](https://github.com/user-attachments/assets/192dbd5a-195b-4e2a-a550-cdf0796b57df)

This looks good. Now I prompt "heart". This gives me exactly the same thing though:

![Image](https://github.com/user-attachments/assets/52e03797-8d4e-4447-99d9-22ba228cfa2e)

When prompting "lung", I get the full lung, which is great:

![Image](https://github.com/user-attachments/assets/726749f0-6a0c-469f-9ed0-1faae4833bee)

When prompting "liver" though, I get pretty much the same segmentation.

![Image](https://github.com/user-attachments/assets/50250528-bc88-4e40-aeb8-268bba55ed74)

Now, switching lung nodule scan provided by you in the examples: When prompting "nodule", I get:

![Image](https://github.com/user-attachments/assets/f61f82ed-20ed-455a-acd1-901c4bf2b7e8)

This again looks desired. When I now try to prompt something different, such as "lung", "aorta" etc., I always get the same result:

![Image](https://github.com/user-attachments/assets/1f6f429f-b165-4071-acb4-10c9731b9ee1)

So I am really not sure, if this behavior is desired, if there is a bug in the code or if I am just not using the network in the desired way. It would be great if you could give some feedback on this.

Thanks so much!

**Comment by liyihao76** [comment]
I have the same problem, it looks like the model doesn't seem to understand the prompts very well, should we use more complex prompts or inference parameters?

**Comment by Victorouledi** [comment]
I though it was due to interactive_infer_image() function (in inference_utils/inference.py) taking as output  the mask embedding which matches the most the prompt text embeddings  no matter the similarity score like following: 

    out_prob = vl_similarity(v_emb, t_emb, temperature=temperature)
    
    matched_id = out_prob.max(0)[1]
    pred_masks_pos = pred_masks[matched_id,:,:]
    pred_class = results['pred_logits'][0][matched_id].max(dim=-1)[1]

I tried to normalize the out_prob values, which were not probabilities, between 0 and 1 and set a threshold, thinking that if one of the predicted mask embeddings didn't have a high enough similarity score with the prompt text embeddings, the model would not make a false prediction.  : 


out_prob = vl_similarity(v_emb, t_emb, temperature=temperature)    
# Apply sigmoid to normalize out_prob to be between 0 and 1
out_prob = torch.sigmoid(out_prob)

threshold = 0.9  
max_prob_value = out_prob.max()

if max_prob_value <= threshold:
    print(f"No match was found between the prompt value {prompts} and the considered image)
    return None 
else:
    # If the probability is sufficient, continue with retrieving the matched_id
    matched_id = out_prob.max(0)[1]


It still doesn't work and it seems that "out_prob = vl_similarity(v_emb, t_emb, temperature=temperature)" does not use the link between the mask embeddings (v_emb) and the label corresponding to the prompt value (t_emb) that the training should have used obviously.

How does the model link different predicted masks with the prompt value that it has seen during the training ?

**Comment by Victorouledi** [comment]
I though it was due to interactive_infer_image() function (in inference_utils/inference.py) taking as output the mask embedding which matches the most the prompt text embeddings no matter the similarity score like following:

out_prob = vl_similarity(v_emb, t_emb, temperature=temperature)

matched_id = out_prob.max(0)[1]
pred_masks_pos = pred_masks[matched_id,:,:]
pred_class = results['pred_logits'][0][matched_id].max(dim=-1)[1]

I tried to normalize the out_prob values, which were not probabilities, between 0 and 1 and set a threshold, thinking that if one of the predicted mask embeddings didn't have a high enough similarity score with the prompt text embeddings, the model would not make a false prediction. :

out_prob = vl_similarity(v_emb, t_emb, temperature=temperature)

out_prob = torch.sigmoid(out_prob)  # Apply sigmoid to normalize out_prob to be between 0 and 1


threshold = 0.9
max_prob_value = out_prob.max()

if max_prob_value <= threshold:
print(f"No match was found between the prompt value {prompts} and the considered image)
return None
else:
matched_id = out_prob.max(0)[1] # If the probability is sufficient, continue with retrieving the matched_id


It still doesn't work and it seems that "out_prob = vl_similarity(v_emb, t_emb, temperature=temperature)" does not use the link between the mask embeddings (v_emb) and the label corresponding to the prompt value (t_emb) that the training should have used obviously.

How does the model link different predicted masks with the prompt value that it has seen during the training ?

**Comment by theodore-zhao** [comment]
Thanks for trying our model and providing the feedback! To reduce the false positive output, there is the check_mask_stats() function in https://github.com/microsoft/BiomedParse/blob/main/inference_utils/output_processing.py. Given the image and predicted mask, the function will output the p-values that the predicted segmentation belongs to the object type in the given modality. The p-values is used to signify false positives, and you can filter out outputs with low p-values (say less than 10e-5). I will put a notebook for that example. 

Thanks!

**Comment by Victorouledi** [comment]
Thanks for your answers @theodore-zhao, once again your work is an invaluable resource !!

**Comment by theodore-zhao** [comment]
We added negative prediction p-value example in inference_example_DICOM.ipynb. Please check it out and hope that works for your scenarios.
